test/Core.Test/Tools/Services/NonAnonymousSendCommandTests.cs

```diff
@@ -0,0 +1,1157 @@
+        }
+        else
+        {
+            // For existing Sends
+            await _sendRepository.Received(1).UpsertAsync(send);
+            Assert.NotEqual(initialDate, send.RevisionDate);
+            await _pushNotificationService.Received(1).PushSyncSendUpdateAsync(send);
+            await _referenceEventService.DidNotReceive().RaiseEventAsync(Arg.Any<ReferenceEvent>());
+        }
+    }
+
+    [Fact]
+    public async Task SaveSendAsync_ExistingSend_Updates()
+    {
+        // Arrange
+        var userId = Guid.NewGuid();
+        var sendId = Guid.NewGuid();
+        var send = new Send
+        {
+            Id = sendId,
+            Type = SendType.Text,
+            UserId = userId,
+            Data = "Some text data"
+        };
+
+        var initialDate = DateTime.UtcNow.AddMinutes(-5);
+        send.RevisionDate = initialDate;
+
+        // Act
+        await _nonAnonymousSendCommand.SaveSendAsync(send);
+
+        // Assert
+        // Verify validation was called
+        await _sendValidationService.Received(1).ValidateUserCanSaveAsync(userId, send);
+
+        // Verify repository was called with updated send
+        await _sendRepository.Received(1).UpsertAsync(send);
+
+        // Check that the revision date was updated
+        Assert.NotEqual(initialDate, send.RevisionDate);
+
+        // Verify push notification was sent for the update
+        await _pushNotificationService.Received(1).PushSyncSendUpdateAsync(send);
+
+        // Verify no reference event was raised (only happens for new sends)
+        await _referenceEventService.DidNotReceive().RaiseEventAsync(Arg.Any<ReferenceEvent>());
+    }
+
+    [Fact]
+    public async Task SaveFileSendAsync_TextType_ThrowsBadRequest()
+    {
+        // Arrange
+        var userId = Guid.NewGuid();
+        var send = new Send
+        {
+            Id = Guid.NewGuid(),
+            Type = SendType.Text, // Text type instead of File
+            UserId = userId
+        };
+        var fileData = new SendFileData();
+        var fileLength = 1024L; // 1KB
+
+        // Act & Assert
+        var exception = await Assert.ThrowsAsync<BadRequestException>(() =>
+            _nonAnonymousSendCommand.SaveFileSendAsync(send, fileData, fileLength));
+
+        Assert.Contains("not of type \"file\"", exception.Message);
+
+        // Verify no further methods were called
+        await _sendValidationService.DidNotReceive().StorageRemainingForSendAsync(Arg.Any<Send>());
+        await _sendRepository.DidNotReceive().CreateAsync(Arg.Any<Send>());
+        await _sendRepository.DidNotReceive().UpsertAsync(Arg.Any<Send>());
+        await _sendFileStorageService.DidNotReceive().GetSendFileUploadUrlAsync(Arg.Any<Send>(), Arg.Any<string>());
+        await _pushNotificationService.DidNotReceive().PushSyncSendCreateAsync(Arg.Any<Send>());
+        await _pushNotificationService.DidNotReceive().PushSyncSendUpdateAsync(Arg.Any<Send>());
```

Comment:
🎨 This can be done more elegantly by configuring the object to throw when these methods are called.

Reply:
Fixed

Reply:
I should have been more precise. If your substitutes **throw by default**, then you can override just those methods that you should be calling to provide your desired behaviors. This was easy to set up in Moq; it's not clear how to set it up in NSubstitute. The goal is to eliminate these blocks completely.

As written, this is wrong. The `throw` calls aren't assertions, they're behaviors. As such, in order to be effective they need to be in the "arrange" stage of your test.

If NSubstitute can't provide the "unless I say otherwise, throw" behavior, then your original assertions are preferable.

Reply:
Reverted to the previous code.